### ISSUE TITLE
Type Mismatch Safety

### ISSUE THREAD

**Issue by jtronge** [rsmpi/rsmpi#133]
While working on error handling related to issue #130, I started thinking about other areas where we might run into some safety issues with current RSMPI code. One of these is with type mismatches, which, depending on the MPI implementation, may or may not be caught. Here's an example program that will compile today:

```
use mpi;
use mpi::traits::*;

fn main() {
    let universe = mpi::initialize().unwrap();
    let world = universe.world();
    let size = world.size();
    let rank = world.rank();
    let mut recv: [i32; 4] = [0; 4];
    let send: [f32; 4] = [1.1, 2.3, 4.4, 5.8];

    if (rank % 2) == 0 {
        world.process_at_rank((rank + 1) % size).send(&send);
        println!("{:?}", send);
    } else {
        world.process_at_rank((rank + size - 1) % size).receive_into(&mut recv[..]);
        println!("{:?}", recv);
    }
}

```
So basically I'm just sending 4 `i32`s and then trying to receive 4 `f32`s. Since these buffers should be of the same size, the MPI implementation (in my case OpenMPI) doesn't care that these buffers are of different types. So, this little example will compile and then run producing garbage output. In a simple program like this it's pretty easy to see the problem here, but I can imagine that in more complex codes this will become a more serious safety issue. I think that most MPI implementations will fail if the buffers don't have the exact same size, so this would probably catch errors like sending `f32`s and trying to receive `f64`s. But there's still the possibility that an error like this could occur, and the program will continue on as if nothing happened.

Looking at section 3.3 of the standard (https://www.mpi-forum.org/docs/mpi-4.0/mpi40-report.pdf), they say that programs like the one above will cause undefined behavior, however they don't require that MPI implementations check this. In particular, under the _Advice to implementors_ in section 3.3.2, they say "The current definition does not require messages to carry data type information."

So it looks like we might have to do something in RSMPI in order to catch this at compile time or at runtime.

**Comment by jtronge** [comment]
This is related to issue #54 and the discussion in PR #51.

**Comment by jedbrown** [comment]
Indeed, this is a safety concern. FWIW, it is valid to receive into a buffer that's "too long". This is used frequently for irregular data. You get the status and inquire using `MPI_Get_count`. We use this in a few dozen places in PETSc, for example. (Those could likely be consolidated, but the point is that it's not super fringe usage.)

**Comment by jtronge** [comment]
Thanks, I didn't realize that. That would seem to make this even more of a concern, since a mismatch could occur in more places without being caught.

**Comment by jedbrown** [comment]
I don't know if point-to-point can even be meaningfully "safe" without significant performance consequences. But point-to-point is usually a building block for higher level collectives. And I could imagine making typed bands containing a communicator and a statically-specified type. Creating such a band would be formally collective, but only for error-checking that the types are consistent. Anyway, I'm mostly interested in making collectives safe and making it possible to write domain-relevant collectives in Rust (with or without a limited amount of unsafe).

**Comment by jtronge** [comment]
Hmmn, so I wonder if we should make the point-to-point API unsafe, while leaving a safe version of the collective API? I agree that if we allow general point-to-point operations, then it would be extremely hard to ensure that everything is safe. Recently I was looking at some research that tried using serde serialization to ensure safety, but they had a huge performance hit by doing that.

I like your idea about using "typed bands" for collective operations. I've started thinking about doing something similar with a sort of agreement or promise type, that would specify the order of collectives to be called and the types that will be used. This could be done in a sort of builder API, perhaps something like this:

```
let a = world.agreement()
    .scatter::<f64>()
    .reduce::<f64>()
    .build();
...
a.scatter(...);
...
a.reduce(...);
```
Once this agreement (or maybe pattern is a better word for it) is built, one single collective containing a specification of this pattern could be passed to each process. This would then ensure that all calls are done in the correct order, and panic if the code doesn't do as it promises.

Maybe this type of code could also be applied to certain peer-to-peer cases. There are often specific patterns that might be used in a code, like with always communicating under a specific topology, etc. We might be able to capture these in a similar way.

**Comment by jedbrown** [comment]
That's interesting, but how would you envision the type system enforce that one not get the order backwards
```rust
a.reduce();
a.scatter();
```
or even do this?
```rust
if rank % 2 == 0 {
  a.reduce();
}
```
The agreement type could process at run-time, but that's barely better and it would make the code really brittle in an environment with hierarchical composition because you'd end up with APIs to create the contract that nearly duplicate the code that performs the work on real data.

**Comment by jtronge** [comment]
Yeah, you're right. This would add a lot of extra code that would need to be maintained. If we did this with a runtime check, it's very likely that subtle bugs would only show up in production.

For a compile-time check originally I was thinking about perhaps some sort of hierarchical type, which would be pretty complicated. For instance for the example I had above, you might end up with something like `AgreementScatter<f64, AgreementReduce<f64, AgreementEnd>>`. Then you'd have to write the communication code similar to this:

```
let a = a.reduce(...);
let a = a.scatter(...);
```

Each call would take ownership, perform the required communication and then return the inner agreement value, or some dummy struct (like `AgreementEnd`).

I'm not even sure if that would work, and if it did it still requires a lot of extra code to manage this set up, as you mentioned. 

Recently, though, I've been thinking about a simpler solution. What if we could attach something to each message/collective that indicated the datatype? If we can represent this as a simple integer or maybe a hash, then I don't think this would cause too much overhead. 

As a runtime check, this could still introduce bugs that aren't caught until much later. And perhaps, although out of scope here, maybe this is something that should be done at the level of MPI itself, rather than within RSMPI.

**Comment by jedbrown** [comment]
I'm not concerned about the bandwidth to send the identifier. The complexity is partly that we'd need a collective sense of what a datatype is. They can be created dynamically and independently, and even include absolute addresses (though I think not in rsmpi). The other issue is that memory can reside on GPUs and sometimes the NIC is connected directly to the GPU so it's not easy to get extra bytes into the message payload. This may be a place where we just can't build safe interfaces without significantly deviating from the MPI model. That said, I think the rules of collective programming usually aren't hard to follow once you understand them, in contrast from say, ownership/borrow-checking in languages other than Rust.

**Comment by jtronge** [comment]
Yes, I'm starting to think that this might not be possible without a major performance hit, unless something changes in MPI.

I looked further at how the MPI standard defines derived datatypes. In section 5.1, they describe a _type map_ which is a sequence of `(type, displacement)`, where type is a primitive. They also describe the _type signature_ which is just a sequence of the primitive types. I'm thinking that the _type signature_ could easily be hashed and used as a runtime type check. Of course this probably not something that rsmpi would be able to do efficiently, and there are also other problems that you mentioned with this.

**Comment by jedbrown** [comment]
As far as the MPI standard is concerned, types need only be compatible. So you can send an array of count=6 doubles and receive count=2 copies of a derived type with 3 doubles. Indeed, the send and receive types could each use absolute addresses and be non-contiguous. Also note that one can post a receive that is bigger than the incoming message (and use status to determine how much was actually received). So creating a hash or other compact encoding that can be checked is more subtle than it appears at first glance. If we want rsmpi to be sound in the way Rust users expect, we may have to make point-to-point unsafe, and make sure we're building good tools for higher level abstractions (mostly collectives, perhaps nonblocking).

**Comment by jtronge** [comment]
Oh, I see your point. Given the way the MPI standard is today, perhaps it's best to make these methods unsafe.

For your example with sending and receiving doubles, while I understand that this is how the standard currently works, I wonder if that scenario is ever really desired? Wouldn't it be better to always require that the two ends of communication always agree on the exact derived/predefined type?

I think this would still allow you to a post a receive that's larger than what was actually sent, as long as what was sent contains only elements of the requested datatype.

**Comment by jedbrown** [comment]
Consider a halo region for a PDE solver. You can use a derived type to offload packing/unpacking, but the send will be from the "interior" of the owner domain and received into the halo region, so those will be different derived types. You could say that because MPI implementations generally aren't very optimized in their packing, callers should always pack explicitly, but that either requires extra kernel launches (on GPUs) or potentially poor interaction with threads on the CPU side. We could eschew the feature from Rust, but that's taking an opinionated stand rather than exposing MPI functionality.

**Comment by hppritcha** [comment]
Good point about the possible matching of a send with a derived type to a receive with a pre-defined type.

**Comment by jtronge** [comment]
Ok, thanks. I think I'm starting to understand the complexity involved here. This definitely won't work as easily as I had thought.

For now, as you noted earlier, the best option is probably to just make point-to-point communication unsafe. Looking back at the discussion in PR #51, I think what they suggested, which is to make all receive operations unsafe, is probably the best way to go.

**Comment by jedbrown** [comment]
Does our definition of safe mean unenforceable dependence on other ranks? In that case, normal and synchronous sends would also be unsafe, while `Isend` could be safe (but waiting on the request would not be). And if that's the case, `Irecv` could also be safe (it's just saying "you can write into this suitably sized buffer") but the wait again would be unsafe. Those rules feel awkward.

**Comment by hppritcha** [comment]
How does rust handle Unix sockets semantics?  One can certainly have a socket open, write data into the socket, then close the socket and if the receiver process at the other end of the socket (or server its running on) dies after the write has complete and the socket closed, what does rust do?

Is there a write up somewhere on how rust safety maps to things like distributed network applications?  I don't see any way a compiler-based safety model could catch all the possible network glitchs that can occur.  Or does rust punt in general on these types of problems with respect to safety.  Or do rusticons approach this by writing additional layers of software on top of the underlying network transport layer to make things more "safe" in a rust sense?

**Comment by jtronge** [comment]
@jedbrown I'm not sure about that. I've been thinking about this in terms of where memory corruption would be possible. Based on that, I believe that synchronous sends would be considered safe, since the memory corruption would only be possible on the receive side, depending on how it interprets the data.

@hppritcha With Unix sockets I think they return a `Result` for writing the data, but as for if the receiver process fails before actually reading it, I don't think there's anything that rust can do or tries to do (see https://doc.rust-lang.org/std/os/unix/net/struct.UnixStream.html). In that case, memory corruption would also only be possible on whichever side is receiving the data.

As far as I know, I don't think Rust has any specific guidelines or models for writing distributed network applications. The main part when working with these types of applications (as well as others) is that all safe API calls are guaranteed to be free from memory-corruption/safety issues.

**Comment by jedbrown** [comment]
This is interesting discussion [`fn read(&mut self, buf: &mut [u8]) -> Result<usize>`](https://doc.rust-lang.org/std/io/trait.Read.html#tymethod.read):
> It is your responsibility to make sure that buf is initialized before calling read. Calling read with an uninitialized buf (of the kind one obtains via [MaybeUninit<T>](https://doc.rust-lang.org/std/mem/union.MaybeUninit.html)) is not safe, and can lead to undefined behavior.

Meanwhile, [`read_to_string`](https://doc.rust-lang.org/std/io/trait.Read.html#method.read_to_string) will error
> If the data in this stream is not valid UTF-8 then an error is returned and `buf` is unchanged.

This implies that it *must* use a scratch buffer and have parsing copy it over after validating the whole thing. I interpret from the above that it's acceptably safe to have an interface that returns `Result` but might deadlock.

This [lovely post](https://manishearth.github.io/blog/2022/08/03/zero-copy-2-zero-copy-all-the-things/) discusses zero-copy deserialization of data structures from buffers. Note that it still needs `fn validate_byte_slice(bytes: &[u8]) -> Result<(), ZeroVecError>`.

I guess we could have MPI receives populate a `MaybeUninit<T>` buffer and another interface to verify that buffer (which maybe could safely be a no-op for types in which all bytes are "valid").
